pkg/operator/machinedeletionhooks/machinedeletionhooks.go

```diff
@@ -0,0 +1,219 @@
+			klog.V(2).Infof("successfully removed the deletion hook from machine %v", machinePendingDeletion.Name)
+			recorder.Eventf("MachineDeletionHook", "successfully removed the deletion hook from machine %v as it doesn't have a member", machinePendingDeletion.Name)
+		}
+		return err
+	}
+	return nil
+}
+
+// addDeletionHookToMasterMachines adds the Machine Deletion Hook to master machines
+func (c *machineDeletionHooksController) addDeletionHookToMasterMachines(ctx context.Context) error {
+	masterMachines, err := c.masterMachineLister.List(c.masterMachineSelector)
+	if err != nil {
+		return err
+	}
+	for _, masterMachine := range masterMachines {
+		if masterMachine.DeletionTimestamp != nil {
+			continue
+		}
+		if hasCompleteMachineDeletionHook(masterMachine) {
+			continue
+		}
+		masterMachineCopy := masterMachine.DeepCopy()
+		if hasPartialMachineDeletionHook(masterMachineCopy) {
+			repairPartialMachineDeletionHook(masterMachineCopy)
+		} else {
+			addPreDrainHook(masterMachineCopy)
+		}
+		if _, err := c.machineClient.Update(ctx, masterMachineCopy, metav1.UpdateOptions{}); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func hasCompleteMachineDeletionHook(machine *machinev1beta1.Machine) bool {
+	for _, hook := range machine.Spec.LifecycleHooks.PreDrain {
+		if hook.Name == ceohelpers.MachineDeletionHookName && hook.Owner == ceohelpers.MachineDeletionHookOwner {
+			return true
+		}
+	}
+	return false
+}
+
+func hasPartialMachineDeletionHook(machine *machinev1beta1.Machine) bool {
+	for _, hook := range machine.Spec.LifecycleHooks.PreDrain {
+		if hook.Name == ceohelpers.MachineDeletionHookName && hook.Owner != ceohelpers.MachineDeletionHookOwner {
+			return true
+		}
+	}
+	return false
+}
+
+func repairPartialMachineDeletionHook(machine *machinev1beta1.Machine) {
+	var newPreDrainHooks []machinev1beta1.LifecycleHook
+
+	for _, hook := range machine.Spec.LifecycleHooks.PreDrain {
+		if hook.Name == ceohelpers.MachineDeletionHookName {
+			continue
+		}
+		newPreDrainHooks = append(newPreDrainHooks, hook)
+	}
+	machine.Spec.LifecycleHooks.PreDrain = newPreDrainHooks
+	addPreDrainHook(machine)
+}
+
+func addPreDrainHook(machine *machinev1beta1.Machine) {
+	machine.Spec.LifecycleHooks.PreDrain = append(machine.Spec.LifecycleHooks.PreDrain, machinev1beta1.LifecycleHook{Name: ceohelpers.MachineDeletionHookName, Owner: ceohelpers.MachineDeletionHookOwner})
+}
+
+func removeMachineDeletionHook(machine *machinev1beta1.Machine) {
+	var newPreDrainHooks []machinev1beta1.LifecycleHook
+	for _, hook := range machine.Spec.LifecycleHooks.PreDrain {
+		if hook.Name == ceohelpers.MachineDeletionHookName {
+			continue
+		}
+		newPreDrainHooks = append(newPreDrainHooks, hook)
+	}
+	machine.Spec.LifecycleHooks.PreDrain = newPreDrainHooks
+}
+
+func indexMachinesByNodeInternalIP(machines []*machinev1beta1.Machine) map[string]*machinev1beta1.Machine {
```

Comment:
how is this different than IndexMachinesByNodeInternalIP from the helpers?

Reply:
cruft, will reuse once https://github.com/openshift/cluster-etcd-operator/pull/780 merges